### PR TITLE
avoid collecting unused threads

### PR DESCRIPTION
**Pull request by trapacska**

### CODE REVIEW

File: tools/filterwriter/filterwriter.go [hunk 1]
```diff
@@ -20,17 +20,19 @@ type Writer struct {
 	writer  io.Writer
 	secrets [][][]byte
 
-	chunk []byte
-	store [][]byte
-	mux   sync.Mutex
+	chunk     []byte
+	store     [][]byte
+	mux       sync.Mutex
+	lastTimer *time.Timer
```
Comment (godrei) [inline tools/filterwriter/filterwriter.go]: why not just `timer`?

File: tools/filterwriter/filterwriter.go [hunk 2]
```diff
@@ -20,17 +20,19 @@ type Writer struct {
...
 }
 
 // New ...
 func New(secrets []string, target io.Writer) *Writer {
-	return &Writer{
+	writer := &Writer{
```
Comment (godrei) [inline tools/filterwriter/filterwriter.go]: what is the reason for this change? it leads to an extra line, but i do not see any advantage, am i missing something?

File: tools/filterwriter/filterwriter.go [hunk 3]
```diff
@@ -20,17 +20,19 @@ type Writer struct {
...
 		writer:  target,
 		secrets: secretsByteList(secrets),
 	}
+	return writer
```
Comment (BirmacherAkos) [inline tools/filterwriter/filterwriter.go]: Is there a reason for creating a writer and return it instead of return the new writer object immediately?

Comment (trapacska) [reply tools/filterwriter/filterwriter.go]: very good catch, it was previously but I forgot to convert it back, thanks!

File: tools/filterwriter/filterwriter.go [hunk 4]
```diff
@@ -52,7 +54,12 @@ func (w *Writer) Write(p []byte) (int, error) {
 	w.chunk = chunk
 	if len(chunk) > 0 {
 		// we have remaining bytes, do not swallow them
-		time.AfterFunc(10*time.Second, func() {
+		if w.lastTimer != nil {
+			w.lastTimer.Stop()
+			w.lastTimer.C = nil
+			w.lastTimer = nil
```
Comment (godrei) [inline tools/filterwriter/filterwriter.go]: i think no need for this line. the new assignment will have the same effect (e.g.: there is no remaining reference for the object and garbage collector will clear it)

File: tools/tools.go [hunk 5]
```diff
@@ -370,6 +370,15 @@ func EnvmanRun(envstorePth, workDirPth string, cmdArgs []string, timeout time.Du
 	cmd.AppendEnv("PWD=" + workDirPth)
 
 	err := cmd.Start()
+
+	// flush the writer anyway if the process is finished
+	if configs.IsSecretFiltering {
```
Comment (godrei) [inline tools/tools.go]: for me it seems the last write will do the flush if there is any leftover chunk, am i missing something?

Comment (trapacska) [reply tools/tools.go]: When the command closes, the timers set will be destroyed so it won't be able to call flush, or maybe it will be flushed in the wrong time.

This way we can make sure that if the command closes, the remaining relevant info will be written there.

File: tools/filterwriter/filterwriter.go (hunk 1 above)
Comment (trapacska) [reply tools/filterwriter/filterwriter.go]: could be

File: tools/filterwriter/filterwriter.go (hunk 4 above)
Comment (trapacska) [reply tools/filterwriter/filterwriter.go]: true
